File: 백준/Gold/1561. 놀이 공원/놀이 공원.cc
#include <algorithm>
#include <iostream>
using namespace std;

int a[10000];

int main() {
    long long n, m;
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        cin >> a[i];
    }

    long long left = 0, right = *min_element(a, a + m) * (n + 1);
    long long ans = 0;
    while (left <= right) {
        long long mid = (left + right) / 2;
        long long cnt = 0;
        for (int j = 0; j < m; j++) {
            cnt += mid / a[j] + 1;
        }

        if (cnt >= n) {
            long long cnt2 = 0;
            for (int i = m - 1; i >= 0; i--) {
                cnt2 += mid / a[i];
                if (mid % a[i] != 0) {
                    cnt2++;
                }
            }
            for (int i = 0; i < m; i++) {
                if (n == cnt2) {
                    ans = i + 1;
                    break;
                }
                if (mid % a[i] == 0) {
                    cnt2++;
                }
                if (n == cnt2) {
                    ans = i + 1;
                    break;
                }
            }
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    cout << ans;
}

File: 백준/Gold/1561. 놀이 공원/놀이 공원.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		int n, m;
		var st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());

		long[] a = new long[m];
		st = new StringTokenizer(br.readLine());
		long ma = 0;
		for (int i = 0; i < m; i++) {
			a[i] = Integer.parseInt(st.nextToken());
			ma = Math.max(ma, a[i]);
		}

		long left = 0, right = (n - 1) * ma;
		long t = 0;

		while (left <= right) {
			long mid = (left + right) / 2;

			long sum = 0;
			for (int i = 0; i < m; i++) {
				sum += mid / a[i] + 1;
			}

			if (sum >= n) {
				right = mid - 1;
				t = mid;
			} else {
				left = mid + 1;
			}
		}

		long s = 0;
		if (t != 0) {
			for (int i = 0; i < m; i++) {
				s += (t - 1) / a[i] + 1;
			}
		}
		for (int i = 0; i < m; i++) {
			if (t % a[i] == 0) {
				s++;
				if (s == n) {
					System.out.println(i + 1);
					return;
				}
			}
		}
	}
}